faculty/session.py

```diff
@@ -26,20 +30,116 @@
+    def _hash_profile(profile):
+        return str(hash(profile))
+
+    def __getitem__(self, profile):
+        return self.tokens[self._hash_profile(profile)]
+
+    def __setitem__(self, profile, access_token):
+        self.tokens[self._hash_profile(profile)] = access_token
+
+    def get(self, profile):
+        try:
+            return self[profile]
+        except KeyError:
+            return None
+
+
+class AccessTokenStoreSchema(Schema):
+    tokens = fields.Dict(
+        keys=fields.String(),
+        values=fields.Nested(AccessTokenSchema),
+        required=True,
+    )
+
+    @post_load
+    def make_access_token_store(self, data):
+        return AccessTokenStore(**data)
+
+
+def _is_valid_access_token(access_token_or_none):
+    if access_token_or_none is None:
+        return False
+    else:
+        return access_token_or_none.expires_at >= datetime.now(tz=pytz.utc)
+
+
 class MemoryAccessTokenCache(object):
     def __init__(self):
-        self._store = {}
+        self._store = AccessTokenStore()
 
     def get(self, profile):
         access_token = self._store.get(profile)
-        utc_now = datetime.now(tz=pytz.utc)
-        if access_token is None or access_token.expires_at < utc_now:
-            return None
+        return access_token if _is_valid_access_token(access_token) else None
+
+    def add(self, profile, access_token):
+        self._store[profile] = access_token
+
+
+def _ensure_directory_exists(path, mode):
+    try:
+        os.makedirs(path, mode=mode)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            # Directory already exists
+            pass
         else:
-            return access_token
+            raise
+
+
+class FileSystemAccessTokenCache(object):
+    def __init__(self, cache_path):
+        self.cache_path = str(cache_path)
+        self._store = None
+
+    def _load_from_disk(self):
+        try:
+            with open(self.cache_path, "r") as fp:
+                data = json.load(fp)
+            self._store = AccessTokenStoreSchema().load(data)
+        except IOError as e:
+            if e.errno == errno.ENOENT:
+                # File does not exist - initialise empty store
+                self._store = AccessTokenStore()
+            else:
+                raise
+        except (ValueError, ValidationError):
+            # File is of invalid format - reset with empty store
+            self._store = AccessTokenStore()
+
+    def _persist_to_disk(self):
+        dirname = os.path.dirname(self.cache_path)
+        _ensure_directory_exists(dirname, mode=0o700)
+        data = AccessTokenStoreSchema().dump(self._store)
+        with open(self.cache_path, "w") as fp:
+            json.dump(data, fp, separators=(",", ":"))
+
+    def get(self, profile):
+        if self._store is None:
+            self._load_from_disk()
+        access_token = self._store.get(profile)
+        return access_token if _is_valid_access_token(access_token) else None
```

Comment:
```suggestion
        if _is_valid_access_token(access_token):
            return access_token
```
(although I'm also happy with explicitly returning `None` in the failure case).

Reply:
Yeah I have a preference for explicitly returning None in this case.